File: 79-word-search/word-search.java
class Solution {

    boolean visited[][];
    public boolean exist(char[][] board, String word) {
        int rows = board.length;
        int cols = board[0].length;
        visited = new boolean[rows][cols];
        
        if(word.length() == 0){return false;}
        for(int r=0; r<rows; r++){
            for(int c=0; c<cols; c++){
                if(word.charAt(0) == board[r][c]){
                    if(search(r, c, word, board, 0)){return true;}
                }
            }
        }
        return false;
    }

    public boolean search(int r, int c, String word, char[][] board,int k){

        if(k == word.length()){return true;}

        if(r >= board.length || r < 0 || c >= board[0].length || c < 0 || visited[r][c] || board[r][c] != word.charAt(k)){return false;}

        visited[r][c] = true;

        if(board[r][c] == word.charAt(k)){
            if (search(r+1, c, word, board, k+1) || search(r-1, c, word, board, k+1)
            || search(r, c+1, word, board, k+1) || search(r, c-1, word, board, k+1) ){
                return true;
            }
        }

        visited[r][c] = false;
        return false;
    }
}

File: 79-word-search/word-search.py
class Solution:
    def exist(self, board: List[List[str]], word: str) -> bool:
        rows, cols = len(board), len(board[0])
        visited = set()

        def dfs(r,c, i):
            if i == len(word):
                return True
            if r >= rows or r < 0 or c>=cols or c<0 or (r,c) in visited or board[r][c] != word[i]:
                return False
            
            visited.add((r,c))
            res = (dfs(r+1, c, i+1) or dfs(r, c+1, i+1) or dfs(r, c-1, i+1) or dfs(r-1, c, i+1))
            visited.remove((r,c))
            return res

        for row in range(rows):
            for col in range(cols):
                if dfs(row, col, 0):
                    return True
        return False
